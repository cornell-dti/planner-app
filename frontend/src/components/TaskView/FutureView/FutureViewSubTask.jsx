// @flow strict

import React from 'react';
import type { Node } from 'react';
import { connect } from 'react-redux';
import { Icon } from 'semantic-ui-react';
import styles from './FutureViewTask.css';
import {
  editSubTask as editSubTaskAction,
  removeSubTask as removeSubTaskAction,
} from '../../../store/actions';
import type { PartialSubTask, SubTask } from '../../../store/store-types';
import type { EditSubTaskAction, RemoveSubTaskAction } from '../../../store/action-types';
import CheckBox from '../../UI/CheckBox';

type Props = {|
  ...SubTask;
  +mainTaskId: number;
  +mainTaskCompleted: boolean;
  +editSubTask: (
    taskId: number, subtaskId: number, partialSubTask: PartialSubTask,
  ) => EditSubTaskAction;
  +removeSubTask: (taskId: number, subTaskId: number) => RemoveSubTaskAction;
|};

/**
 * The component used to render one subtask in backlog day.
 *
 * @param props the props to render.
 * @return {Node} the rendered element.
 * @constructor
 */
function FutureViewSubTask(props: Props): Node {
  const {
    name, id, mainTaskId, complete, inFocus,
    mainTaskCompleted, editSubTask, removeSubTask,
  } = props;
  return (
    <div className={styles.SubTask}>
      <CheckBox
        className={styles.TaskCheckBox}
        checked={mainTaskCompleted || complete}
        disabled={mainTaskCompleted}
        inverted
        onChange={() => { editSubTask(mainTaskId, id, { complete: !complete }); }}
      />
      <span
        className={styles.TaskText}
        style={(mainTaskCompleted || complete) ? { textDecoration: 'line-through' } : {}}
      >
        {name}
      </span>
      <Icon
        name={inFocus ? 'bookmark' : 'bookmark outline'}
        className={styles.TaskIcon}
        onClick={() => editSubTask(mainTaskId, id, { inFocus: !inFocus })}
      />
      <Icon
        name="delete"
        className={styles.TaskIcon}
        onClick={() => removeSubTask(mainTaskId, id)}
      />
    </div>
  );
}

const ConnectedFutureViewSubTask = connect(
  null,
  { editSubTask: editSubTaskAction, removeSubTask: removeSubTaskAction },
)(FutureViewSubTask);
export default ConnectedFutureViewSubTask;
