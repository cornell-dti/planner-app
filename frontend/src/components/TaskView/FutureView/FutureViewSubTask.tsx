import React, { ReactElement } from 'react';
import { SubTask } from 'common/types/store-types';
import styles from './FutureViewTask.module.scss';
import CheckBox from '../../UI/CheckBox';
import { editSubTask, removeSubTask } from '../../../firebase/actions';
import SamwiseIcon from '../../UI/SamwiseIcon';

type Props = {
  readonly subTask: SubTask | null;
  readonly mainTaskId: string;
  readonly replaceDateForFork: Date | null;
  readonly mainTaskCompleted: boolean;
};

/**
 * The component used to render one subtask in future view day.
 */
function FutureViewSubTask({
  subTask,
  mainTaskId,
  replaceDateForFork,
  mainTaskCompleted,
}: Props): ReactElement | null {
  if (subTask == null) {
    return null;
  }
  const { name, complete, inFocus } = subTask;
  const onCompleteChange = (): void =>
    editSubTask(mainTaskId, subTask.id, replaceDateForFork, { complete: !complete });
  const onFocusChange = (): void =>
    editSubTask(mainTaskId, subTask.id, replaceDateForFork, { inFocus: !inFocus });
  const onRemove = (): void => removeSubTask(mainTaskId, subTask.id, replaceDateForFork);
  return (
    <div className={styles.SubTask}>
      <CheckBox
        className={styles.SubTaskCheckBox}
        checked={mainTaskCompleted || complete}
        disabled={mainTaskCompleted}
        inverted
        onChange={onCompleteChange}
      />
      <span
        className={styles.TaskText}
        style={mainTaskCompleted || complete ? { textDecoration: 'line-through' } : {}}
      >
        {name}
      </span>
      <SamwiseIcon
        iconName={inFocus ? 'pin-dark-filled' : 'pin-dark-outline'}
        onClick={onFocusChange}
        className={styles.TaskIcon}
      />
      <SamwiseIcon iconName="x-dark" className={styles.TaskIcon} onClick={onRemove} />
    </div>
  );
}

const Memoized = React.memo(FutureViewSubTask);
export default Memoized;
