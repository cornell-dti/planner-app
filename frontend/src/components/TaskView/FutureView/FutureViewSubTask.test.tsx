import React from 'react';
import renderer from 'react-test-renderer';
import FutureViewSubTask from './FutureViewSubTask';

it('FutureViewSubTask with null SubTask matches snapshot.', () => {
  const tree = renderer.create(
    <FutureViewSubTask
      mainTaskId="foo"
      subTask={null}
      replaceDateForFork={null}
      mainTaskCompleted
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

const createTest = (inFocus: boolean, complete: boolean, mainTaskCompleted: boolean): void => {
  const testName = `FutureViewSubTask(${inFocus}, ${complete}, ${mainTaskCompleted}) matches snapshot`;
  it(testName, () => {
    const tree = renderer.create(
      <FutureViewSubTask
        mainTaskId="foo"
        subTask={{ id: 'id', order: 1, name: 'foo', inFocus, complete }}
        replaceDateForFork={null}
        mainTaskCompleted={mainTaskCompleted}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
};

createTest(false, false, false);
createTest(false, false, true);
createTest(false, true, false);
createTest(false, true, true);
createTest(true, false, false);
createTest(true, false, true);
createTest(true, true, false);
createTest(true, true, true);
