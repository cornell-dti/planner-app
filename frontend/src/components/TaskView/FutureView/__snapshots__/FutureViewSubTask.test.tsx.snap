// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FutureViewSubTask with null SubTask matches snapshot. 1`] = `null`;

exports[`FutureViewSubTask(false, false, false) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={false}
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
  </label>
  <span
    className="TaskText"
    style={Object {}}
  >
    foo
  </span>
  <span
    title="Pin to Focus"
  >
    <svg
      alt="Pin to Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Pin to Focus"
      width="1em"
    >
      pin-2-dark-outline.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;

exports[`FutureViewSubTask(false, false, true) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={true}
      disabled={true}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
    <span
      className="CheckBoxCheckMark"
    />
  </label>
  <span
    className="TaskText"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
  >
    foo
  </span>
  <span
    title="Pin to Focus"
  >
    <svg
      alt="Pin to Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Pin to Focus"
      width="1em"
    >
      pin-2-dark-outline.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;

exports[`FutureViewSubTask(false, true, false) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={true}
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
    <span
      className="CheckBoxCheckMark"
    />
  </label>
  <span
    className="TaskText"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
  >
    foo
  </span>
  <span
    title="Pin to Focus"
  >
    <svg
      alt="Pin to Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Pin to Focus"
      width="1em"
    >
      pin-2-dark-outline.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;

exports[`FutureViewSubTask(false, true, true) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={true}
      disabled={true}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
    <span
      className="CheckBoxCheckMark"
    />
  </label>
  <span
    className="TaskText"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
  >
    foo
  </span>
  <span
    title="Pin to Focus"
  >
    <svg
      alt="Pin to Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Pin to Focus"
      width="1em"
    >
      pin-2-dark-outline.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;

exports[`FutureViewSubTask(true, false, false) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={false}
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
  </label>
  <span
    className="TaskText"
    style={Object {}}
  >
    foo
  </span>
  <span
    title="Unpin from Focus"
  >
    <svg
      alt="Unpin from Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Unpin from Focus"
      width="1em"
    >
      pin-2-dark-filled.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;

exports[`FutureViewSubTask(true, false, true) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={true}
      disabled={true}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
    <span
      className="CheckBoxCheckMark"
    />
  </label>
  <span
    className="TaskText"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
  >
    foo
  </span>
  <span
    title="Unpin from Focus"
  >
    <svg
      alt="Unpin from Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Unpin from Focus"
      width="1em"
    >
      pin-2-dark-filled.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;

exports[`FutureViewSubTask(true, true, false) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={true}
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
    <span
      className="CheckBoxCheckMark"
    />
  </label>
  <span
    className="TaskText"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
  >
    foo
  </span>
  <span
    title="Unpin from Focus"
  >
    <svg
      alt="Unpin from Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Unpin from Focus"
      width="1em"
    >
      pin-2-dark-filled.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;

exports[`FutureViewSubTask(true, true, true) matches snapshot 1`] = `
<div
  className="SubTask"
>
  <label
    className="TaskCheckBox CheckBox InvertedCheckBox"
  >
    <input
      defaultChecked={true}
      disabled={true}
      onClick={[Function]}
      tabIndex={0}
      type="checkbox"
    />
    <span
      className="CheckBoxCheckMark"
    />
  </label>
  <span
    className="TaskText"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
  >
    foo
  </span>
  <span
    title="Unpin from Focus"
  >
    <svg
      alt="Unpin from Focus"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Unpin from Focus"
      width="1em"
    >
      pin-2-dark-filled.svg
    </svg>
  </span>
  <span
    title="Delete"
  >
    <svg
      alt="Delete"
      className="TaskIcon"
      height="1em"
      onClick={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      title="Delete"
      width="1em"
    >
      XDark.svg
    </svg>
  </span>
</div>
`;
